### ISSUE TITLE
Always resolve timeout on IPv6 only network

### ISSUE THREAD

**Issue by BusyJay** [grpc/grpc#32704]
### What version of gRPC and what language are you using?
1.46.5

### What operating system (Linux, Windows,...) and version?
Linux

### What did you do?
1) Configure an IPv6 only cluster using kubernetes
2) Starts a greeter example and accessing a hostname

### What did you expect to see?
The greeter should work fine.

### What did you see instead?
It timesout. Sets GRPC_VERBOSITY to DEBUG and GRPC_TRACE to all, it gives following logs:
```
(c-ares resolver) request:0x7f84b809e0e0 create_hostbyname_request_locked host:basic-peer port:19209 is_balancer:
0 qtype:AAAA
(c-ares resolver) request:0x7f84b809e0e0 create_hostbyname_request_locked host:basic-peer port:19209 is_balancer:
0 qtype:A
(c-ares resolver) request:0x7f84b809e0e0 ev_driver=0x7f84b804b800 grpc_ares_ev_driver_start_locked. timeout in 
120005 ms
(c-ares resolver) request:0x7f84b809e0e0 ev_driver=0x7f84b804b800. next ares process poll time in 1000 ms
(c-ares resolver) resolver:0x7f84b8063180 Started resolving. pending_request_:0x7f84b809e0e0
(c-ares resolver) request:0x7f84b809e0e0 on_hostbyname_done_locked qtype=AAAA host=basic-peer ARES_SUCCESS
(c-ares resolver) request:0x7f84b809e0e0 c-ares resolver gets a AF_INET6 result: \n  addr: fd00:100:64::cd77\n 
 port: 2379\n  sin6_scope_id: 0\n
(c-ares resolver) request:0x7f84b809e0e0 ev_driver=0x7f84b804b800 on_ares_backup_poll_alarm_locked. driver->shu
tting_down=0. err=\"OK\"
(c-ares resolver) request:0x7f84b809e0e0 ev_driver=0x7f84b804b800 on_ares_backup_poll_alarm_locked; ares_proces
s_fd. fd=c-ares fd: 11
(c-ares resolver) request:0x7f84b809e0e0 ev_driver=0x7f84b804b800. next ares process poll time in 1000 ms
...
```
Some logs are omitted. DNS server will only return IPv6 record, for IPv4 query it will return NXDOMAIN. gRPC wait for both IPv4 and IPv6 records ready, so it will only start connecting remote after resolve timeout, which is 120s by default and is too long for any latency sensitive service.

A way to fix the problem is providing an arg for client to tell grpc it's an IPv6-only network and stop resolving IPv4.

**Comment by sanmai-NL** [comment]
Use e.g., the [Happy Eyeballs algorithm](https://eng.libretexts.org/Bookshelves/Computer_Science/Networks/Book:_An_Introduction_to_Computer_Networks_(Dordal)/08:_IP_version_6/8.12:_Using_IPv6_and_IPv4_Together).

**Comment by yijiem** [comment]
@BusyJay do you see any error message for the IPv4 query such as: `C-ares status is not ARES_SUCCESS qtype=A name=<...> is_balancer=<...>: DNS server returned answer with no data` in your log before the timeout? If it was not there, then it seems strange that the DNS server has responded with NXDOMAIN, but C-ares did not fail the query call with error.

**Comment by BusyJay** [comment]
No, I don't see it. I believe C-ares will keep retrying when seeing NXDOMAIN.

**Comment by BusyJay** [comment]
If enable dual stack, for IPv4 query, CoreDNS returns NOERROR with empty data, then gRPC will return quickly, and there are logs like `C-ares status is not ARES_SUCCESS qtype=A name=<...> is_balancer=<...>: DNS server returned answer with no data`.

**Comment by yashykt** [comment]
Yijie, Alex, this is possibly a real bug. Please take a look.

**Comment by apolcyn** [comment]
In general, c-ares should only be retrying errors that look due to transient unavailability. E.g. like if it can't contact a DNS server. If c-ares actually gets an NXDOMAIN response, it should accept that and won't retry it.

So I'm very skeptical of the explanation that c-ares is retrying NXDOMAIN errrors.

We could determine exactly what's happening here with a couple of logs:
1) run `GRPC_VERBOSITY=DEBUG` and `GRPC_TRACE=cares_resolver,cares_address_sorting` as you did here but paste the complete log
2) capture DNS activity alongside your program with `tcpdump` (e.g. capture all traffic on port 53 to see what queries c-ares is sending and what responses it's getting). Please also paste that here if possible.

**Comment by lobshunter** [comment]
Hi, @apolcyn, I am co-worker of @BusyJay. 

We did some test again and confirmed that c-ares handles NXDOMAIN as expected. The problem is I didn't configure DNS server correctly, so some DNS requests time-outed and retried indefinitely. After correcting the DNS config everything works fine.

Thanks for helping and apologize if any confusion caused.

**Comment by lobshunter** [comment]
Run more tests after since, looks like GRPC will keep trying to resolve domain name after getting a `NOERROR` _empty response_. But tools like `dig` or `nslookup` would stop resolution in such case. This could be a problem in some situations(for our case, a kubernetes IPv6 only environment).

In the screenshot below, GRPC tries to resolve `basic-pd`. It firstly requests A record of `basic-pd.user-py.svc.cluster.local`(because `/etc/resolv.conf` said so), and gets an NOERROR empty response. Then requests `basic-pd.svc.cluster.local` instead of stopping. What causes problem is the request will eventually be forwarded to upstream DNS server(and in our case the server would drop unknown requests, result in resolution timeout).
<img width="1797" alt="Screenshot 2023-03-29 at 18 26 37" src="https://user-images.githubusercontent.com/65525602/228508352-40f5916c-8d3a-4cc4-adf2-a5f525125abe.png">

Is is possible for GRPC to stop resolution after getting an empty response? (Although I didn't find related DNS RFCs). 

---
**Attachment:**
Raw pcap file:
[empty-noerror.gz](https://github.com/grpc/grpc/files/11099632/empty-noerror.gz)

And /etc/resolv.conf of the pod:
```txt
search user-py.svc.cluster.local svc.cluster.local cluster.local testbed-ipv6-7hmvs.svc.cluster.local default.svc.cluster.local
nameserver fd00:100:64::a
options ndots:5
```

**Comment by lobshunter** [comment]
> I think nslookup and dig are not good behavioral comparisons here because I believe those tools do not implement domain search list functionality - rather they just send oneshot DNS queries.

I might disagree with this. `nslookup` does follow the domain search list, but only when it gets an failed response(FAIL = SERVFAIL | NXDOMAIN | REFUSED)


> A quick way to test the libc resolver behavior would be with python. E.g. try something like:

I did it with the snippet in `python:lastest`(python@sha256:5329e75033c4446dc92d702cf8ebbeb63e549d9b83076a776f6753e10817fc3c) container, glibc 2.31. It did stop A record resolution after getting an empty response.

It is true that our DNS had inappropriate config. Using FQDN might not be a good option for kubernetes environment(but that's out of this topic). 
So I am still kind of wondering if c-ares can accept DNS empty response as valid answer.

**Comment by apolcyn** [comment]
Interesting.

> I did it with the snippet in python:lastest(python@sha256:5329e75033c4446dc92d702cf8ebbeb63e549d9b83076a776f6753e10817fc3c) container, glibc 2.31. It did stop A record resolution after getting an empty response.

I think this might be a good question for the c-ares library maintainers. Would you have time to move this to https://github.com/c-ares/c-ares ?

Note, gRPC just uses the `ares_gethostbyname` API for host name resolution:

- IPv6 records are queried in https://github.com/grpc/grpc/blob/656e5f283b9cc5ba0020962d8ddee4c30758d692/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L740
- IPv4 records are queried in https://github.com/grpc/grpc/blob/656e5f283b9cc5ba0020962d8ddee4c30758d692/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L745
